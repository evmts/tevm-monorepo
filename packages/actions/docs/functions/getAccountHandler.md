**@tevm/actions** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > getAccountHandler

# Function: getAccountHandler()

> **getAccountHandler**(`evm`): `GetAccountHandler`

Creates an GetAccountHandler for handling account params with Ethereumjs EVM

## Parameters

▪ **evm**: `EVM`

## Returns

## Source

[packages/actions/src/handlers/getAccountHandler.js:11](https://github.com/evmts/tevm-monorepo/blob/main/packages/actions/src/handlers/getAccountHandler.js#L11)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
