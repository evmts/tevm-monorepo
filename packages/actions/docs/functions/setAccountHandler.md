[**@tevm/actions**](../README.md) • **Docs**

***

[@tevm/actions](../globals.md) / setAccountHandler

# Function: setAccountHandler()

> **setAccountHandler**(`client`, `options`?): [`SetAccountHandler`](../type-aliases/SetAccountHandler.md)

## Parameters

• **client**: `TevmNode`\<`"fork"` \| `"normal"`, `object`\>

• **options?** = `{}`

• **options.throwOnFail?**: `undefined` \| `boolean`

whether to default to throwing or not when errors occur

## Returns

[`SetAccountHandler`](../type-aliases/SetAccountHandler.md)

## Defined in

[packages/actions/src/SetAccount/setAccountHandler.js:17](https://github.com/evmts/tevm-monorepo/blob/main/packages/actions/src/SetAccount/setAccountHandler.js#L17)
