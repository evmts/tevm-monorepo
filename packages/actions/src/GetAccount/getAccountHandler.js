import { createAddress } from '@tevm/address'
import { AccountNotFoundError, ForkError, InternalError } from '@tevm/errors'
import { bytesToHex } from '@tevm/utils'
import { cloneVmWithBlockTag } from '../Call/cloneVmWithBlock.js'
import { getPendingClient } from '../internal/getPendingClient.js'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'
import { validateGetAccountParams } from './validateGetAccountParams.js'

/**
 * Creates an GetAccountHandler for handling account params with Ethereumjs VM
 * @param {import("@tevm/node").TevmNode} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('../GetAccount/GetAccountHandlerType.js').GetAccountHandler}
 */
export const getAccountHandler =
	(client, options = {}) =>
	async ({ throwOnFail = options.throwOnFail ?? true, ...params }) => {
		const vm = await client.getVm()
		/**
		 * @type {Array<import('../GetAccount/TevmGetAccountError.js').TevmGetAccountError>}
		 */
		const errors = validateGetAccountParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, {
				errors,
				address: params.address,
				balance: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				storageRoot: '0x',
				nonce: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				deployedBytecode: '0x',
				/**
				 * @type {`0x${string}`}
				 */
				codeHash: '0x',
				isContract: false,
				isEmpty: true,
			})
		}

		const address = createAddress(params.address)
		try {
			if (params.blockTag === 'pending') {
				const mineResult = await getPendingClient(client)
				if (mineResult.errors) {
					return maybeThrowOnFail(throwOnFail, {
						errors: mineResult.errors,
						address: params.address,
						balance: 0n,
						/**
						 * @type {`0x${string}`}
						 */
						storageRoot: '0x',
						nonce: 0n,
						/**
						 * @type {`0x${string}`}
						 */
						deployedBytecode: '0x',
						/**
						 * @type {`0x${string}`}
						 */
						codeHash: '0x',
						isContract: false,
						isEmpty: true,
					})
				}
				return getAccountHandler(mineResult.pendingClient, options)({ throwOnFail, ...params, blockTag: 'latest' })
			}
			if (params.blockTag !== 'latest' && params.blockTag !== undefined) {
				const block = await vm.blockchain.getBlockByTag(params.blockTag)
				const clonedVm = await cloneVmWithBlockTag(client, block)
				if (clonedVm instanceof ForkError || clonedVm instanceof InternalError) {
					return maybeThrowOnFail(throwOnFail, {
						errors: [clonedVm],
						address: params.address,
						balance: 0n,
						/**
						 * @type {`0x${string}`}
						 */
						storageRoot: '0x',
						nonce: 0n,
						/**
						 * @type {`0x${string}`}
						 */
						deployedBytecode: '0x',
						/**
						 * @type {`0x${string}`}
						 */
						codeHash: '0x',
						isContract: false,
						isEmpty: true,
					})
				}
				return getAccountHandler(
					{ ...client, getVm: () => Promise.resolve(clonedVm) },
					options,
				)({ throwOnFail, ...params, blockTag: 'latest' })
			}
			const res = await vm.stateManager.getAccount(address)
			if (!res) {
				return maybeThrowOnFail(throwOnFail, {
					address: params.address,
					balance: 0n,
					/**
					 * @type {`0x${string}`}
					 */
					storageRoot: '0x',
					nonce: 0n,
					/**
					 * @type {`0x${string}`}
					 */
					deployedBytecode: '0x',
					errors: [new AccountNotFoundError(`account ${params.address} not found`)],
					/**
					 * @type {`0x${string}`}
					 */
					codeHash: '0x',
					isContract: false,
					isEmpty: true,
				})
			}
			const code = res?.codeHash !== undefined ? bytesToHex(await vm.stateManager.getContractCode(address)) : '0x'

			return {
				// TODO some of these fields are not in the api and should be added to @tevm/actions
				address: params.address,
				balance: res.balance,
				codeHash: bytesToHex(res.codeHash),
				isContract: res.isContract(),
				isEmpty: res.isEmpty(),
				deployedBytecode: code,
				nonce: res.nonce,
				storageRoot: bytesToHex(res.storageRoot),
				...(params.returnStorage
					? {
							storage: Object.fromEntries(
								Object.entries(await vm.stateManager.dumpStorage(address)).map(([key, value]) => [
									`0x${key}`,
									/** @type {import('../common/Hex.js').Hex}*/ (value),
								]),
							),
						}
					: {}),
			}
		} catch (e) {
			let err = e
			// TODO this isn't clean
			// What we are doing here is checking if we threw a known error or not
			if (typeof e !== 'object' || e === null || !('_tag' in e)) {
				err = new InternalError('UnexpectedError in getAccountHandler', { cause: /** @type {any}*/ (e) })
			}
			errors.push(/** @type any*/ (err))
			return maybeThrowOnFail(throwOnFail, {
				errors,
				address: params.address,
				balance: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				storageRoot: '0x',
				/**
				 * @type {`0x${string}`}
				 */
				codeHash: '0x',
				nonce: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				deployedBytecode: '0x',
				isContract: false,
				isEmpty: true,
			})
		}
	}
