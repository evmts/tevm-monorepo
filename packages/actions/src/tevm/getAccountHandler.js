import { createError } from './createError.js'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import { EthjsAddress } from '@tevm/utils'
import { bytesToHex, hexToBytes } from '@tevm/utils'
import { validateGetAccountParams } from '@tevm/zod'

/**
 * Creates an GetAccountHandler for handling account params with Ethereumjs VM
 * @param {Pick<import('@tevm/base-client').BaseClient, 'getVm'>} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('@tevm/actions-types').GetAccountHandler}
 */
export const getAccountHandler =
	(client, options = {}) =>
	async ({ throwOnFail = options.throwOnFail ?? true, ...params }) => {
		const vm = await client.getVm()
		/**
		 * @type {Array<import('@tevm/errors').GetAccountError>}
		 */
		const errors = validateGetAccountParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, {
				errors,
				address: params.address,
				balance: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				storageRoot: '0x',
				nonce: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				deployedBytecode: '0x',
				/**
				 * @type {`0x${string}`}
				 */
				codeHash: '0x',
				isContract: false,
				isEmpty: true,
			})
		}

		const address = new EthjsAddress(hexToBytes(params.address))
		try {
			const res = await vm.stateManager.getAccount(address)
			if (!res) {
				return maybeThrowOnFail(throwOnFail, {
					address: params.address,
					balance: 0n,
					/**
					 * @type {`0x${string}`}
					 */
					storageRoot: '0x',
					nonce: 0n,
					/**
					 * @type {`0x${string}`}
					 */
					deployedBytecode: '0x',
					errors: [
						createError(
							'AccountNotFoundError',
							`account ${params.address} not found`,
						),
					],
					/**
					 * @type {`0x${string}`}
					 */
					codeHash: '0x',
					isContract: false,
					isEmpty: true,
				})
			}
			const code =
				res?.codeHash !== undefined
					? bytesToHex(await vm.stateManager.getContractCode(address))
					: '0x'
			return {
				// TODO some of these fields are not in the api and should be added to @tevm/actions-types
				address: params.address,
				balance: res.balance,
				codeHash: bytesToHex(res.codeHash),
				isContract: res.isContract(),
				isEmpty: res.isEmpty(),
				deployedBytecode: code,
				nonce: res.nonce,
				storageRoot: bytesToHex(res.storageRoot),
			}
		} catch (e) {
			errors.push(
				createError(
					'UnexpectedError',
					typeof e === 'string'
						? e
						: e instanceof Error
						? e.message
						: 'unknown error',
				),
			)
			return maybeThrowOnFail(throwOnFail, {
				errors,
				address: params.address,
				balance: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				storageRoot: '0x',
				/**
				 * @type {`0x${string}`}
				 */
				codeHash: '0x',
				nonce: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				deployedBytecode: '0x',
				isContract: false,
				isEmpty: true,
			})
		}
	}
