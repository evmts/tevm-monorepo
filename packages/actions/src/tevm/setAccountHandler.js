import { createError } from './createError.js'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import { EthjsAccount, EthjsAddress } from '@tevm/utils'
import { hexToBytes, keccak256 } from '@tevm/utils'
import { validateSetAccountParams } from '@tevm/zod'

/**
 * Creates an SetAccountHandler for handling account params with Ethereumjs EVM
 * @param {Pick<import('@tevm/base-client').BaseClient, 'vm'>} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('@tevm/actions-types').SetAccountHandler}
 */
export const setAccountHandler = (client, options = {}) => async (params) => {
	const { throwOnFail = options.throwOnFail ?? true } = params
	/**
	 * @type {Array<import('@tevm/errors').SetAccountError>}
	 */
	const errors = validateSetAccountParams(params)
	if (errors.length > 0) {
		return maybeThrowOnFail(throwOnFail, { errors })
	}

	const address = new EthjsAddress(hexToBytes(params.address))
	try {
		await client.vm.stateManager.putAccount(
			address,
			new EthjsAccount(
				params.nonce,
				params.balance,
				params.storageRoot && hexToBytes(params.storageRoot),
				params.deployedBytecode &&
					hexToBytes(keccak256(params.deployedBytecode)),
			),
		)
		if (params.deployedBytecode) {
			await client.vm.stateManager.putContractCode(
				address,
				hexToBytes(params.deployedBytecode),
			)
		}
		await client.vm.stateManager.checkpoint()
		await client.vm.stateManager.commit()
		// TODO offer way of setting contract storage with evm.stateManager.putContractStorage
		return {}
	} catch (e) {
		errors.push(
			createError(
				'UnexpectedError',
				typeof e === 'string'
					? e
					: e instanceof Error
					? e.message
					: 'unknown error',
			),
		)
		return maybeThrowOnFail(throwOnFail, { errors })
	}
}
