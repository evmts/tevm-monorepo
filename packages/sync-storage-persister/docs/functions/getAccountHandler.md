**@tevm/actions** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > getAccountHandler

# Function: getAccountHandler()

> **getAccountHandler**(`client`, `options`?): `GetAccountHandler`

Creates an GetAccountHandler for handling account params with Ethereumjs VM

## Parameters

▪ **client**: `Pick`\<`object`, `"vm"`\>

▪ **options?**: `object`= `{}`

▪ **options.throwOnFail?**: `undefined` \| `boolean`

whether to default to throwing or not when errors occur

## Returns

## Source

[packages/actions/src/tevm/getAccountHandler.js:15](https://github.com/evmts/tevm-monorepo/blob/main/packages/actions/src/tevm/getAccountHandler.js#L15)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
