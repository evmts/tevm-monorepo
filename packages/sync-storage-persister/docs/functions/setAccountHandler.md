**@tevm/actions** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > setAccountHandler

# Function: setAccountHandler()

> **setAccountHandler**(`client`, `options`?): `SetAccountHandler`

Creates an SetAccountHandler for handling account params with Ethereumjs EVM

## Parameters

▪ **client**: `Pick`\<`object`, `"vm"`\>

▪ **options?**: `object`= `{}`

▪ **options.throwOnFail?**: `undefined` \| `boolean`

whether to default to throwing or not when errors occur

## Returns

## Source

[packages/actions/src/tevm/setAccountHandler.js:14](https://github.com/evmts/tevm-monorepo/blob/main/packages/actions/src/tevm/setAccountHandler.js#L14)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
