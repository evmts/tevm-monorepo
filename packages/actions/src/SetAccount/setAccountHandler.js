import { createAddress } from '@tevm/address'
import { AccountNotFoundError, InternalError } from '@tevm/errors'
import { EthjsAccount } from '@tevm/utils'
import { hexToBytes, keccak256 } from '@tevm/utils'
import { getAccountHandler } from '../GetAccount/getAccountHandler.js'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'
import { validateSetAccountParams } from './validateSetAccountParams.js'

/**
 * Creates an SetAccountHandler for handling account params with Ethereumjs EVM
 * @param {import("@tevm/node").TevmNode} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('./SetAccountHandlerType.js').SetAccountHandler}
 */
export const setAccountHandler =
	(client, options = {}) =>
	async (params) => {
		const { throwOnFail = options.throwOnFail ?? true } = params
		/**
		 * @type {Array<import('../SetAccount/TevmSetAccountError.js').TevmSetAccountError>}
		 */
		const errors = validateSetAccountParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, { errors })
		}

		const address = createAddress(params.address)

		/**
		 * @type {Array<Promise<any>>}
		 */
		const promises = []
		try {
			const vm = await client.getVm()

			// check if account exists
			const account = await getAccountHandler(client)({ ...params, throwOnFail: false })
			if (account.errors?.length && !(account.errors[0] instanceof AccountNotFoundError)) {
				client.logger.error('there was an unexpected error getting account', account.errors)
				throw account.errors.length > 1 ? new AggregateError(account.errors) : account.errors[0]
			}

			promises.push(
				vm.stateManager.putAccount(
					address,
					new EthjsAccount(
						params.nonce ?? account?.nonce,
						params.balance ?? account?.nonce,
						(params.storageRoot && hexToBytes(params.storageRoot)) ??
							(account?.storageRoot !== undefined && account?.storageRoot !== '0x'
								? hexToBytes(account.storageRoot)
								: undefined),
						(params.deployedBytecode && hexToBytes(keccak256(params.deployedBytecode))) ??
							(account?.deployedBytecode !== undefined ? hexToBytes(keccak256(account.deployedBytecode)) : undefined),
					),
				),
			)
			if (params.deployedBytecode) {
				promises.push(vm.stateManager.putContractCode(address, hexToBytes(params.deployedBytecode)))
			}
			// state clears state first stateDiff does not
			if (params.state) {
				await vm.stateManager.clearContractStorage(address)
			}
			const state = params.state ?? params.stateDiff
			if (state) {
				for (const [key, value] of Object.entries(state)) {
					promises.push(
						vm.stateManager.putContractStorage(
							address,
							hexToBytes(/** @type {import('@tevm/utils').Hex}*/ (key), { size: 32 }),
							hexToBytes(value),
						),
					)
				}
			}
			const results = await Promise.allSettled(promises)
			for (const result of results) {
				if (result.status === 'rejected') {
					errors.push(new InternalError('Unable to put storage', { cause: result.reason }))
				}
			}

			if (errors.length > 0) {
				return maybeThrowOnFail(throwOnFail, { errors })
			}
			await vm.stateManager.checkpoint()
			await vm.stateManager.commit(false)

			return {}
		} catch (e) {
			errors.push(new InternalError('Unexpected error setting account', { cause: e }))
			return maybeThrowOnFail(throwOnFail, { errors })
		}
	}
