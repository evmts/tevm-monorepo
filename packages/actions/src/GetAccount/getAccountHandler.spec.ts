import { AccountNotFoundError, InvalidAddressError } from '@tevm/errors'
import { createTevmNode } from '@tevm/node'
import { TestERC20 } from '@tevm/test-utils'
import { numberToHex } from 'viem'
import { describe, expect, it } from 'vitest'
import { setAccountHandler } from '../SetAccount/setAccountHandler.js'
import { getAccountHandler } from './getAccountHandler.js'

const contract = TestERC20.withAddress(`0x${'3'.repeat(40)}`)

describe('getAccount', () => {
	it('should get an account from evm', async () => {
		const client = createTevmNode()
		const res = await setAccountHandler(client)({
			address: contract.address,
			deployedBytecode: contract.deployedBytecode,
			balance: 420n,
			nonce: 69n,
		})
		expect(res.errors).toBeUndefined()
		const account = await getAccountHandler(client)({
			address: contract.address,
		})
		expect(account?.balance).toBe(420n)
		expect(account?.nonce).toBe(69n)
		expect(account.deployedBytecode).toBe(contract.deployedBytecode)
	})

	it('should validate params', async () => {
		const client = createTevmNode()
		const res = await setAccountHandler(client)({
			// @ts-expect-error
			address: 'not an address',
			throwOnFail: false,
		})
		expect(res.errors).toMatchSnapshot()
	})

	it('should handle account not found', async () => {
		const client = createTevmNode()
		const account = await getAccountHandler(client)({
			throwOnFail: false,
			address: `0x${'4'.repeat(40)}`,
		})
		expect(account?.errors?.[0]).toBeInstanceOf(AccountNotFoundError)
		expect(account?.balance).toBe(0n)
		expect(account?.nonce).toBe(0n)
		expect(account?.deployedBytecode).toBe('0x')
		expect(account?.isContract).toBe(false)
		expect(account?.isEmpty).toBe(true)
	})

	it('should handle internal errors gracefully', async () => {
		const client = createTevmNode()
		// Simulate an internal error by passing invalid parameters to getAccountHandler
		const invalidParams = { throwOnFail: false }
		const account = await getAccountHandler(client)(invalidParams as any)
		expect(account?.errors?.[0]).toBeInstanceOf(InvalidAddressError)
		expect(account?.balance).toBe(0n)
		expect(account?.nonce).toBe(0n)
		expect(account?.deployedBytecode).toBe('0x')
		expect(account?.isContract).toBe(false)
		expect(account?.isEmpty).toBe(true)
	})

	it('should return storage if requested', async () => {
		const client = createTevmNode()
		const state = {
			[numberToHex(0, { size: 32 })]: numberToHex(420, { size: 2 }),
			[numberToHex(2, { size: 32 })]: numberToHex(420, { size: 2 }),
		}
		await setAccountHandler(client)({
			address: contract.address,
			deployedBytecode: contract.deployedBytecode,
			balance: 420n,
			nonce: 69n,
			state,
		})
		const account = await getAccountHandler(client)({
			address: contract.address,
			returnStorage: true,
		})
		expect(account?.storage).toBeDefined()
		Object.entries(state).forEach(([key, value]) => {
			expect(account?.storage?.[key as any]).toBe(value)
		})
		expect(account?.storage).toMatchSnapshot()
	})

	it('should handle getAccount unexpectedly throwing', async () => {
		const client = createTevmNode()
		const vm = await client.getVm()
		vm.stateManager.getAccount = () => {
			throw new Error('Unexpected error')
		}
		const account = await getAccountHandler(client)({
			address: contract.address,
			throwOnFail: false,
		})
		expect(account.errors).toBeDefined()
		expect(account.errors).toMatchSnapshot()
	})
})
