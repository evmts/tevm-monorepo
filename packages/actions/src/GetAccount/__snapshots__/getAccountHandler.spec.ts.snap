// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getAccount should validate params 1`] = `
[
  [InvalidAddressError: Invalid Address not an address

Docs: https://tevm.sh/reference/tevm/errors/classes/invalidaddresserror/
Version: 1.1.0.next-73],
]
`;

exports[`getAccount should return storage if requested 1`] = `
{
  "0x0000000000000000000000000000000000000000000000000000000000000000": "0x01a4",
  "0x0000000000000000000000000000000000000000000000000000000000000002": "0x01a4",
}
`;

exports[`getAccount should handle getAccount unexpectedly throwing 1`] = `
[
  [InternalError: UnexpectedError in getAccountHandler

Docs: https://tevm.sh/reference/tevm/errors/classes/internalerror/
Details: Unexpected error
Version: 1.1.0.next-73],
]
`;
